applications/OptimizationApplication/custom_utilities/response/max_overhang_response_utils.cpp

```diff
@@ -0,0 +1,274 @@
+#include "utilities/openmp_utils.h"
+
+// Application includes
+#include "custom_utilities/geometrical/model_part_utils.h"
+#include "optimization_application_variables.h"
+
+// Include base h
+#include "max_overhang_response_utils.h"
+
+namespace Kratos
+{
+
+double MaxOverhangAngleResponseUtils::CalculateValue(
+    const std::vector<ModelPart*>& rModelParts, 
+    const Parameters ResponseSettings)
+{
+    double value = 0.0;    
+    for (auto p_model_part : rModelParts) {
+        const double local_condition_value = block_for_each<SumReduction<double>>(p_model_part->Conditions(), [&](const auto& rCondition) {
+            return CalculateConditionValue(rCondition,ResponseSettings);
+        });
+        value += p_model_part->GetCommunicator().GetDataCommunicator().SumAll(local_condition_value);
+    }
+    
+    double total_area = 0.0;
+    for (auto p_model_part : rModelParts) {
+        const double local_condition_area = block_for_each<SumReduction<double>>(p_model_part->Conditions(), [&](const auto& rCondition) {
+            return rCondition.GetGeometry().Area();
+        });
+        total_area += p_model_part->GetCommunicator().GetDataCommunicator().SumAll(local_condition_area);
+    }
+
+    return value/total_area;
+}
+
+double MaxOverhangAngleResponseUtils::CalculateConditionValue(
+    const Condition& rCondition,
+    const Parameters& rResponseSettings)
+{
+    array_3d print_direction = rResponseSettings["print_direction"].GetVector();
+    KRATOS_ERROR_IF(!(norm_2(print_direction) > std::numeric_limits<double>::epsilon()))
+    << "MaxOverhangAngleResponseUtils:print_direction: "<<print_direction<<" has size close to zero\n";    
+    print_direction /= norm_2(print_direction);
+
+    const double max_angle = rResponseSettings["max_angle"].GetDouble();
+    KRATOS_ERROR_IF(!(max_angle >= 0.0 && max_angle <= 90.0))
+    << "MaxOverhangAngleResponseUtils: max_angle: "<<max_angle<<" should be in range [0,90]\n";
+    double sin_max_angle = std::sin(max_angle * Globals::Pi / 180);
+
+
+    const double heaviside_beta = rResponseSettings["heaviside_beta"].GetDouble();
+    KRATOS_ERROR_IF(std::signbit(heaviside_beta))
+    << "MaxOverhangAngleResponseUtils: heaviside_beta: "<<heaviside_beta<<" should be positive\n";
+
+    const double penalty_factor = rResponseSettings["penalty_factor"].GetDouble();
+    KRATOS_ERROR_IF(std::signbit(penalty_factor))
+    << "MaxOverhangAngleResponseUtils: penalty_factor: "<<penalty_factor<<" should be positive\n";    
+
+    const array_3d local_coords = ZeroVector(3);
+    const array_3d& normal = rCondition.GetGeometry().UnitNormal(local_coords);
+    const double area = rCondition.GetGeometry().Area();
+    const double ratio = -inner_prod(print_direction, normal)/sin_max_angle;
+    double pow_val = -2.0 * heaviside_beta * (ratio-1); 
+    std::clamp(pow_val, -700.0, 700.0);
+
+    return area * (1.0/(1+std::exp(pow_val))) * std::pow(ratio,penalty_factor);
```

Comment:
since we are allowing `pow_val` in the range `[-700, 700]`, did you check what happens if `std::exp(700)`? is it gonna give `nan`?